File: app/src/main/java/com/hiteshsahu/awesome_gallery/view/FullScreenGalleryActivity.java
package com.hiteshsahu.awesome_gallery.view;

import android.content.ContentUris;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AppCompatActivity;
import android.widget.Toast;

import com.hiteshsahu.awesome_gallery.BuildConfig;
import com.hiteshsahu.awesome_gallery.R;
import com.hiteshsahu.awesome_gallery.modal.CenterRepository;
import com.hiteshsahu.awesome_gallery.view.adapter.TouchImageAdapter;
import com.hiteshsahu.awesome_gallery.view.widget.ExtendedViewPager;

import java.io.File;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class FullScreenGalleryActivity extends AppCompatActivity {

    public static final String IMAGE_POSITION = "PagerPosition";

    @BindView(R.id.view_pager)
    ExtendedViewPager mViewPager;
    private TouchImageAdapter galleryAdaqpter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_full_screen_gallery);
        ButterKnife.bind(this);

//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
//            getWindow().setEnterTransition(new Explode());
//            getWindow().setExitTransition(new Explode());
//            getWindow().setSharedElementEnterTransition(new ChangeImageTransform());
//            getWindow().setSharedElementExitTransition(new ChangeImageTransform());
//            getWindow().setSharedElementEnterTransition(TransitionInflater.from(this)
//                    .inflateTransition(R.transition.curve));
//        }

        int currentPagePosition = getIntent().getIntExtra(IMAGE_POSITION, 0);
        galleryAdaqpter = new TouchImageAdapter();
        mViewPager.setAdapter(galleryAdaqpter);
        mViewPager.setCurrentItem(currentPagePosition);

//
//            rootView.findViewById(R.id.info).setOnClickListener(new OnClickListener() {
//
//                @Override
//                public void onClick(View v) {
//
////                    // 1. Instantiate an AlertDialog.Builder with its
////                    // constructor
////                    AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
////
////                    builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
////                        public void onClick(DialogInterface dialog, int id) {
////
////                        }
////                    });
////                    builder.setNegativeButton("Done", new DialogInterface.OnClickListener() {
////                        public void onClick(DialogInterface dialog, int id) {
////                            // User cancelled the dialog
////                        }
////                    });
////
////                    StringBuilder str = new StringBuilder();
////
////                    // 2. Chain together various setter methods to set the
////                    // dialog characteristics
////
////                    // 3. Get the AlertDialog from create()
////
////                    // There are multiple ways to get a Metadata object for
////                    // a file
////
////                    //
////                    // SCENARIO 1: UNKNOWN FILE TYPE
////                    //
////                    // This is the most generic approach. It will
////                    // transparently determine the file type and invoke the
////                    // appropriate
////                    // readers. In most cases, this is the most appropriate
////                    // usage. This will handle JPEG, TIFF, GIF, BMP and RAW
////                    // (CRW/CR2/NEF/RW2/ORF) files and extract whatever
////                    // metadata is available and understood.
////
////                    //
////
////                    File file = getImageFile();
////
////                    try {
////                        Metadata metadata = ImageMetadataReader.readMetadata(file);
////
////                        for (Directory directory : metadata.getDirectories()) {
////
////                            //
////                            // Each Directory stores values in Tag objects
////                            //
////                            for (Tag tag : directory.getTags()) {
////                                System.out.println(tag);
////
////                                str.append(tag.toString());
////                                str.append("\n");
////
////                            }
////
////                            //
////                            // Each Directory may also contain error
////                            // messages
////                            //
////                            if (directory.hasErrors()) {
////                                for (String error : directory.getErrors()) {
////                                    System.err.println("ERROR: " + error);
////                                }
////                            }
////
////                        }
////
////                    } catch (ImageProcessingException e) {
////                        // handle exception
////                    } catch (IOException e) {
////                        // handle exception
////                    }
////
////                    builder.setMessage(str.toString()).setTitle("Image Info");
////
////                    AlertDialog dialog = builder.create();
////
////                    dialog.show();
//
//                }
//            });
//


    }

    @OnClick(R.id.canvas)
    public void editThisItem() {
        Intent editIntent = new Intent(Intent.ACTION_EDIT);
        editIntent.setDataAndType(Uri.fromFile(
                new File(CenterRepository.getInstance().getImageCollection().
                        getImageAt((mViewPager.getCurrentItem())).getImagePath())), "image/*");
        editIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        startActivity(Intent.createChooser(editIntent, null));

    }

    @OnClick(R.id.share)
    public void shareThisItem() {
        Intent share = new Intent(Intent.ACTION_SEND);
        share.setType("image/*");
        share.putExtra(Intent.EXTRA_STREAM,
                FileProvider.getUriForFile(FullScreenGalleryActivity.this,
                        BuildConfig.APPLICATION_ID + ".provider",
                        new File(CenterRepository.getInstance().getImageCollection().
                                getImageAt((mViewPager.getCurrentItem())).getImagePath())));
        startActivity(Intent.createChooser(share, "Share Image"));
    }


    @OnClick(R.id.delete)
    public void deleteThisFile() {

        // Set up the projection (we only need the ID)
        String[] projection = {MediaStore.Images.Media._ID};

        // Match on the file path
        String selection = MediaStore.Images.Media.DATA + " = ?";
        String[] selectionArgs = new String[]{CenterRepository.getInstance().getImageCollection().
                getImageAt((mViewPager.getCurrentItem())).getImagePath()
        };

        // Query for the ID of the media matching the file path
        Cursor imageCursor = getContentResolver()
                .query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        projection,
                        selection,
                        selectionArgs,
                        null);

        if (imageCursor != null) {

            if (imageCursor.moveToFirst()) {
                // We found the ID. Deleting the item via the content
                // provider will also remove the file
                long imageID =
                        imageCursor.getLong(imageCursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID));
                Uri deleteUri =
                        ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                                imageID);
                getContentResolver().delete(deleteUri, null, null);

                CenterRepository.getInstance().getImageCollection().
                        getListOfImages().remove(mViewPager.getCurrentItem());

                galleryAdaqpter.notifyDataSetChanged();

                sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE,
                        Uri.fromFile(new File(CenterRepository.getInstance().getImageCollection().
                                getImageAt((mViewPager.getCurrentItem())).getImagePath()))));

                Toast.makeText(getApplicationContext(), "Deleted", Toast.LENGTH_SHORT).show();

            }

            imageCursor.close();
        }
    }
}

File: app/src/main/java/com/hiteshsahu/awesome_gallery/view/FullScreenGalleryActivity.kt
package com.hiteshsahu.awesome_gallery.view

import android.annotation.SuppressLint
import android.content.ContentUris
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.view.animation.AnimationUtils
import android.widget.TextView
import android.widget.Toast
import android.widget.ViewSwitcher
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import androidx.viewpager2.widget.ViewPager2
import com.hiteshsahu.awesome_gallery.BuildConfig.APPLICATION_ID
import com.hiteshsahu.awesome_gallery.modal.CenterRepository
import com.hiteshsahu.awesome_gallery.view.adapter.GalleryPagerAdapter
import com.hiteshsahu.awesome_gallery.view.adapter.MyAdapter
import com.hiteshsahu.awesome_gallery.view.widget.ParallaxPagerTransformer
import kotlinx.android.synthetic.main.activity_full_screen_gallery.*
import java.io.File


/**
 * Activity to display image full screen with Pinch to Zoom option
 */
class FullScreenGalleryActivity : AppCompatActivity() {

    private var galleryAdapter: GalleryPagerAdapter? = null
    private var fullScreenMode = false;

    @SuppressLint("RestrictedApi")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.hiteshsahu.awesome_gallery.R.layout.activity_full_screen_gallery)

        /*    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getWindow().setEnterTransition(new Explode());
                getWindow().setExitTransition(new Explode());
                getWindow().setSharedElementEnterTransition(new ChangeImageTransform());
                getWindow().setSharedElementExitTransition(new ChangeImageTransform());
                getWindow().setSharedElementEnterTransition(TransitionInflater.from(this)
                        .inflateTransition(R.transition.curve));
            }*/

        val currentPagePosition = intent.getIntExtra(IMAGE_POSITION, 0)



        // On click
        mViewPager.setOnItemClickListener {

            // your code
            if (fullScreenMode) {
                backFab.visibility = View.VISIBLE
                window.addFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                window.clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);

            } else {
                backFab.visibility = View.GONE
                window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                window.clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
            }
            fullScreenMode = !fullScreenMode
        }
        // set adapter
        mViewPager.setPageTransformer(false, ParallaxPagerTransformer(com.hiteshsahu.awesome_gallery.R.id.touchImageView));
        galleryAdapter = GalleryPagerAdapter()
        mViewPager.adapter = galleryAdapter
        mViewPager.currentItem = currentPagePosition



//        // View Pager 2
//        val myAdpater = MyAdapter(this)
//        myViewPager2.orientation = ViewPager2.ORIENTATION_VERTICAL
//        myViewPager2.adapter = myAdpater
//        myViewPager2.currentItem = currentPagePosition
//        myViewPager2.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
//            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
//                super.onPageScrolled(position, positionOffset, positionOffsetPixels)
//            }
//
//            override fun onPageSelected(position: Int) {
//                super.onPageSelected(position)
//                pagePosition.setText("" + position + "/" + galleryAdapter!!.count)
//            }
//
//            override fun onPageScrollStateChanged(state: Int) {
//                super.onPageScrollStateChanged(state)
//            }
//        })


        //Page index
        pagePosition.setFactory(object : ViewSwitcher.ViewFactory {

            override fun makeView(): View {
                val pageIndex = TextView(this@FullScreenGalleryActivity)
                pageIndex.gravity = Gravity.CENTER
                pageIndex.textSize = 15f
                pageIndex.setTextColor(Color.WHITE)
                return pageIndex
            }
        })

        val slideInUp = AnimationUtils.loadAnimation(this, com.hiteshsahu.awesome_gallery.R.anim.slide_in_up)
        val slideOutDown = AnimationUtils.loadAnimation(this, com.hiteshsahu.awesome_gallery.R.anim.slide_out_up)

        pagePosition.inAnimation = slideInUp
        pagePosition.outAnimation = slideOutDown

        mViewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {}
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                pagePosition.setText("" + position + "/" + galleryAdapter!!.count)
            }
        })

        pagePosition.setText("" + currentPagePosition + "/" + galleryAdapter!!.count)


        // edit image
        editImage.setOnClickListener {
            editThisItem(CenterRepository.instance.imageCollection.getImageAt(mViewPager.currentItem).imagePath)
        }

        // share image
        shareImage.setOnClickListener {
            shareThisItem(CenterRepository.instance.imageCollection
                    .getImageAt(mViewPager.currentItem)
                    .imagePath)
        }

        //delete Image
        deleteImage.setOnClickListener {
            deleteThisFile()
        }

    }

    private fun editThisItem(imagePath: String) {
        val editIntent = Intent(Intent.ACTION_EDIT)
        editIntent.type = "image/*"
        editIntent.putExtra(Intent.EXTRA_STREAM,
                FileProvider.getUriForFile(this@FullScreenGalleryActivity,
                        APPLICATION_ID + ".provider",
                        File(imagePath)))

        editIntent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        startActivity(Intent.createChooser(editIntent, null))

    }

    private fun shareThisItem(filePath: String) {
        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.type = "image/*"
        shareIntent.putExtra(Intent.EXTRA_STREAM,
                FileProvider.getUriForFile(this@FullScreenGalleryActivity,
                        APPLICATION_ID + ".provider",
                        File(filePath)))
        startActivity(Intent.createChooser(shareIntent, "Share Image"))
    }


    private fun deleteThisFile() {

        // Set up the projection (we only need the ID)
        val projection = arrayOf(MediaStore.Images.Media._ID)

        // Match on the file path
        val selection = MediaStore.Images.Media.DATA + " = ?"
        val selectionArgs = arrayOf(CenterRepository.instance.imageCollection.getImageAt(mViewPager.currentItem).imagePath)

        // Query for the ID of the media matching the file path
        val imageCursor = contentResolver
                .query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        projection,
                        selection,
                        selectionArgs, null)

        if (imageCursor != null) {

            if (imageCursor.moveToFirst()) {
                // We found the ID. Deleting the item via the content
                // provider will also remove the file
                val imageID = imageCursor.getLong(imageCursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID))
                val deleteUri = ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        imageID)
                contentResolver.delete(deleteUri, null, null)

                CenterRepository.instance.imageCollection.listOfImages.removeAt(mViewPager.currentItem)

                galleryAdapter!!.notifyDataSetChanged()

                sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE,
                        Uri.fromFile(File(CenterRepository.instance.imageCollection.getImageAt(mViewPager.getCurrentItem()).imagePath))))

                Toast.makeText(applicationContext, "Deleted", Toast.LENGTH_SHORT).show()

            }

            imageCursor.close()
        }
    }

    companion object {
        const val IMAGE_POSITION = "PagerPosition"
    }
}

